pyriemann/estimation.py

```diff
@@ -651,3 +651,81 @@ def transform(self, X):
             covmats[ii] = shrunk_covariance(x, self.shrinkage)
 
         return covmats
+
+
+class BlockCovariances(BaseEstimator, TransformerMixin):
+    """Estimation of block covariance matrix.
+    Perform a block covariance matrix estimation for each given trial. The
+    resulting matrices are block diagonal matrices.
+
+    The blocks on the diagonal are calculated as individual covariance
+    matrices for a subset of channels using the given the estimator.
+    Varying block sized possible by passing a list to allow incorporation
+    of different modalities with different number of channels (e.g. EEG,
+    ECoG, LFP, EMG) with their own respective covariance matrices.
+
+    Parameters
+    ----------
+    estimator : string (default: 'scm')
+        covariance matrix estimator. For regularization consider 'lwf' or 'oas'
+        For a complete list of estimator, see `utils.covariance`.
+    block_size : list or int
+        Sizes of individual blocks given as int for same-size block or list for
+        varying block sizes.
+
+    Notes
+    -----
+    .. versionadded:: 0.2.8
+
+    See Also
+    --------
+    Covariances
+    """
+
+    def __init__(self, block_size, estimator='scm'):
+        """Init."""
+        self.estimator = estimator
+        self.block_size = block_size
+
+    def fit(self, X, y=None):
+        """Fit.
+
+        Do nothing. For compatibility purpose.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_matrices, n_channels, n_samples)
+            ndarray of matrices.
+        y : ndarray shape (n_matrices,)
+            labels corresponding to each trial, not used.
+
+        Returns
+        -------
+        self : Covariances instance
+            The Covariances instance.
+        """
+        return self
+
+    def transform(self, X):
+        """Estimate block covariance matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_matrices, n_channels, n_samples)
+            ndarray of matrices.
+
+        Returns
+        -------
+        covmats : ndarray, shape (n_matrices, n_channels, n_channels)
+            ndarray of covariance matrices for each trials.
+        """
+        n_matrices, n_channels, n_times = X.shape
+
+        if isinstance(self.block_size, int):
+            n_blocks = n_channels // self.block_size
+            blocks = [self.block_size for b in range(n_blocks)]
+
+        else:
```

Comment:
```suggestion
        elif isinstance(self.block_size, list):
```
and code must raise an error otherwise.

Reply:
What if people pass the blocks as ndarrays for whatever reason? Is there a way to check if it's array like?

Reply:
I just added check for ndarray. If anyone passes any other array-like, it's on them I guess.

Reply:
I don't see the check for ndarray.
And can you raise an error when it is neither a `int` nor a `list`/`ndarray`?